cmd/state/internal/cmdtree/ppm.go

```diff
@@ -2,128 +2,134 @@ package cmdtree
 
 import (
 	"fmt"
+	"os"
+	"os/exec"
 	"strings"
 
+	"github.com/ActiveState/cli/internal/analytics"
 	"github.com/ActiveState/cli/internal/captain"
+	"github.com/ActiveState/cli/internal/errs"
 	"github.com/ActiveState/cli/internal/locale"
+	"github.com/ActiveState/cli/internal/primer"
+	"github.com/ActiveState/cli/pkg/projectfile"
 )
 
-func printSuggestion(ppmIntent, newCommand, docLink string) error {
-	fmt.Println(locale.Tr("ppm_print_suggestion", ppmIntent, newCommand, docLink))
+func printSuggestion(prime *primer.Values, ppmIntent, newCommand, docLink string) error {
+	prime.Output().Print(locale.Tr("ppm_print_suggestion", ppmIntent, newCommand, docLink))
 	return nil
 }
 
-func printDefault() error {
-	fmt.Println(strings.TrimSpace(locale.T("ppm_header_message")))
+func printDefault(prime *primer.Values) error {
+	prime.Output().Print(strings.TrimSpace(locale.T("ppm_header_message")))
 	return nil
 }
 
-func printMain() error {
-	fmt.Println(locale.T("ppm_print_main"))
+func printMain(prime *primer.Values) error {
+	prime.Output().Print(locale.T("ppm_print_main"))
 	return nil
 }
 
-func newPpmCommand() *captain.Command {
+func newPpmCommand(prime *primer.Values) *captain.Command {
 	rootCmd := captain.NewHiddenShimCommand(
 		"_ppm",
 		nil, nil,
 		func(_ *captain.Command, args []string) error {
 			for _, arg := range args {
 				if arg == "--version" {
-					return printDefault()
+					return printDefault(prime)
 				}
 			}
-			return printMain()
+			return shim(prime, "ppm", "packages", "ppm_print_forward", args...)
```

Comment:
This `shim` stuff only works for printing out basic messages when invoking a command. This new stuff is getting much more advanced. For one we'll need a runner, and for another we need explicitly typed functions for eg. `ppk packages`.

The point of cmdtree is to establish the main command structure, it should not have any command logic, that's what the runner is for.

Reply:
Agreed. I've moved the shimming to a runner. I'm not sure if interpreted the typed functions comment correctly so please let me know.